compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFusePackIntoLoop.cpp

```diff
@@ -18,37 +19,54 @@ namespace mlir::iree_compiler::AMDAIE {
 
 namespace {
 
-/// A utility function specific to this pass which, given a value, would
-/// traverse the def-chain till it either finds a tensor.extract_slice op or a
-/// BlockArgument.
+FailureOr<Operation *> getParentBeforeLoop(BlockArgument blkArg) {
+  Operation *parent = blkArg.getOwner()->getParentOp();
+  LoopLikeOpInterface loop = dyn_cast<LoopLikeOpInterface>(parent);
+  if (!loop) return failure();
+  Operation *operandParent = loop.getTiedLoopInit(blkArg)->getOwner();
+  return operandParent;
+}
+
+/// A utility function specific to this pass which, given a value `operand`,
+/// traverses the def-chain till it finds a tensor.extract_slice. The 2 cases
+/// where it successfully finds and returns an extract_slice (SLICE) are:
+///
+/// Case 1)
+/// pack -> block arg -> SLICE  -> pack -> pack -> pack -> operand
+///                                ^^^^^^^^^^^^^^^^^^^^
+///                                any number (>= 0) of trailing packs
+///
+/// Case 2)
+/// pack -> SLICE -> pack -> pack -> pack -> operand
+///                  ^^^^^^^^^^^^^^^^^^^^
+///                  any number (>= 0) of trailing packs
+///
 static FailureOr<tensor::ExtractSliceOp> getTensorExtractSliceDefiningOp(
     Value operand) {
-  while (Operation *defOp = operand.getDefiningOp()) {
-    auto sliceOp = dyn_cast_if_present<tensor::ExtractSliceOp>(defOp);
-    if (sliceOp) {
-      // The producer of sliceOp should be a pack op.
-      if (isa_and_present<tensor::PackOp>(
-              sliceOp.getSource().getDefiningOp())) {
-        return sliceOp;
-      }
-      if (isa<BlockArgument>(sliceOp.getSource())) {
-        auto blkArg = dyn_cast<BlockArgument>(sliceOp.getSource());
-        for (Value blkOperand :
-             blkArg.getOwner()->getParentOp()->getOperands()) {
-          if (isa_and_present<tensor::PackOp>(blkOperand.getDefiningOp())) {
-            return sliceOp;
-          }
-        }
-      }
-      break;
-    }
-    // We perform further traversal only if we have tensor.pack op in the
-    // def-chain.
-    if (!isa<tensor::PackOp>(defOp)) {
-      break;
-    }
-    operand = defOp->getOperand(0);
+  // roll back through all the packs immediately preceding `operand`.
+  while (operand.getDefiningOp() &&
+         isa<tensor::PackOp>(operand.getDefiningOp())) {
```

Comment:
I think you can use `while (isa_and_present<tensor::PackOp>(operand.getDefiningOp()))`.

Reply:
done